### PR TITLE
Added archives

### PR DESCRIPTION
**Pull request by umbreak**
Fixes https://github.com/BlueBrain/nexus/issues/720

Most relevant items:
- Added schema validation (/schemas/archive.json)
- Added cache (ArchiveCache.scala)
- Added surface API (Archives.scala)
- Added routes (ArchiveRoutes.scala)

### CODE REVIEW

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  /**
+    * Routes for archives. Those routes should get triggered after the following segment have been consumed:
+    * {prefix}/archives/{org}/{project}. E.g.: v1/archives/myorg/myproject
+    */
+  def routes: Route =
+    concat(
+      // Create an archive when id is not provided in the Uri (POST)
+      (post & pathEndOrSingleSlash) {
+        operationName(s"/${config.http.prefix}/archives/{}/{}") {
+          (hasPermission(write) & projectNotDeprecated) {
```
[the remark this answers is not in the document]
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: or maybe `archives/write` is a better fit, thoughts?

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            entity(as[Json]) { source =>
+              complete(archives.create(source).value.runWithStatus(Created))
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: Missing `Location` header with 303 see other.

Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: That's on purpose. I wanted this API to be as similar as the rest of the API. As when you create a file in the system, you can fetch it back using the same endpoint as the PUT but with a GET, here you can use the `self` to fetch either the resource or it's binary content, depending on content negotiation HTTP headers.

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  /**
+    * Routes for archives when the id is specified.
+    * This route should get triggered after the following segment have been consumed:
+    * {prefix}/archives/{org}/{project}/{id}. E.g.: v1/archives/myorg/myproject/myArchive
+    */
+  def routes(id: AbsoluteIri): Route = {
+    val resId = Id(project.ref, id)
+    concat(
+      // Create archive
+      (put & pathEndOrSingleSlash) {
```
[the remark this answers is not in the document]
Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: I think it is better to use a compatible API. I don't think POST / Redirect / Get is needed

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: we can add `archives/write`...but then that will have to be added to IAM configurations, ADMIN configurations and so on.

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/archives/Archive.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  def toTarIgnoreNotFound[F[_]: Effect](implicit fetchResource: FetchResource[F, ArchiveSource]): F[AkkaSource] =
+    values.toList.map(fetchResource(_).value).sequence.map(results => TarFlow.write(results.flatten))
+
+  def toTar[F[_]: Effect](implicit fetchResource: FetchResource[F, ArchiveSource]): OptionT[F, AkkaSource] =
+    values.toList.map(fetchResource(_)).sequence.map(TarFlow.write(_))
+}
+
+object Archive {
+
+  val write: Permission = Permission.unsafe("resources/write")
```
[the remark this answers is not in the document]
Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/archives/Archive.scala]: For the routes.

File: src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: I think this is important, an archive represents a future download, we should discriminate the permissions required.

[the remark this answers is not in the document]
Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/kg/routes/ArchiveRoutes.scala]: Done

File: src/main/resources/app.conf [hunk 5]
```diff
@@ -434,4 +434,22 @@ app {
...
+    cache-ask-timeout = 5 seconds
+    cache-ask-timeout = ${?ARCHIVES_CACHE_ASK_TIMEOUT}
+
+    # the time resource is kept in the archive cache before being invalidated
+    cache-invalidate-after = 5 hours
+    cache-invalidate-after = ${?ARCHIVES_CACHE_INVALIDATE_AFTER}
+
+    # the maximum number of resources that can be contain in the archive
+    max-resources = 1000
+    max-resources = ${?archives_MAX_RESOURCES}
```
Comment (bogdanromanx) [inline src/main/resources/app.conf]: typo here

**Pull request merged**